pkg/instance/handler_test.go

```diff
@@ -0,0 +1,141 @@
+package instance
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/dhis2-sre/im-manager/pkg/model"
+	"github.com/dhis2-sre/im-user/swagger/sdk/models"
+	"github.com/gin-gonic/gin"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"gorm.io/gorm"
+)
+
+func TestHandler_List_ServiceError(t *testing.T) {
+	groups := []*models.Group{
+		{
+			Name:     "name",
+			Hostname: "hostname",
+		},
+	}
+
+	service := &mockInstanceService{}
+	errorMessage := "some error"
+	service.
+		On("FindInstances", groups, false).
+		Return(nil, errors.New(errorMessage))
+
+	handler := NewHandler(nil, service, nil)
+
+	w := httptest.NewRecorder()
+	c, _ := gin.CreateTestContext(w)
+	c.Set("user", &models.User{Groups: groups})
+
+	handler.List(c)
+
+	assert.Empty(t, w.Body.Bytes())
+	assert.NotEmpty(t, c.Errors)
+
+	service.AssertExpectations(t)
+}
+
+func TestHandler_List(t *testing.T) {
+	groups := []*models.Group{
+		{
+			Name:     "name",
+			Hostname: "hostname",
+		},
+	}
+
+	instances := []*model.Instance{
+		{
+			Model:     gorm.Model{ID: 1},
+			Name:      "some name",
+			GroupName: groups[0].Name,
+		},
+	}
+
+	service := &mockInstanceService{}
+	service.
+		On("FindInstances", groups, false).
+		Return(instances, nil)
+
+	handler := NewHandler(nil, service, nil)
+
+	w := httptest.NewRecorder()
+	c, _ := gin.CreateTestContext(w)
+	c.Set("user", &models.User{Groups: groups})
+
+	handler.List(c)
+
+	assert.Empty(t, c.Errors)
+	assert.Equal(t, http.StatusOK, w.Code)
+	var body []GroupWithInstances
+	err := json.Unmarshal(w.Body.Bytes(), &body)
+	require.NoError(t, err)
+	assert.Equal(t, groupsWithInstances(groups, instances), body)
```

Comment:
If you are using this function in your assertion you are not testing it. Maybe we can setup a struct or a raw string with what we expect. A little duplication in tests that with regards to setup is not a problem to me. Testing more without assuming to much about the implementation is more important IMHO. This gives you more confidence and freedom in refactoring 😄

Reply:
So you could either do

```go
    assert.Empty(t, c.Errors)
    assert.Equal(t, http.StatusOK, w.Code)
    var actual []GroupWithInstances
    err := json.Unmarshal(w.Body.Bytes(), &actual)
    require.NoError(t, err)
    expected := []GroupWithInstances{
        {
            Name:     "name",
            Hostname: "hostname",
            Instances: []*model.Instance{
                {
                    Model:     gorm.Model{ID: 1},
                    Name:      "some name",
                    GroupName: "name",
                },
            },
        },
    }
    assert.Equal(t, expected, actual)
```

or

```go
assert.Equal(t, `[{"Name":"name","Hostname":"hostname","Instances":[{"ID":1,"CreatedAt":"0001-01-01T00:00:00Z","UpdatedAt":"0001-01-01T00:00:00Z","DeletedAt":null,"UserID":0,"Name":"some name","GroupName":"name","StackName":"","requiredParameters":null,"optionalParameters":null,"DeployLog":"","Preset":false,"PresetID":0}]}]`, w.Body.String())
```

the latter has the benefit of not having to do the unmarshaling. Since the JSON is not pretty printed its not easy to read and in case the test fails the diff will not be as well.

So I would opt for the first suggestion.

I assume we are asserting some JSON body regularly, so we could write a helper so the test reads nicely.

`assertJSON(t, body, expected)`

we probably also need a struct to unmarshal into.